File: uber/uber_algo/LC56MergeIntervals.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        if not intervals:
            return []
        answer = []
        intervals = sorted(intervals, key=lambda x:x.start)
        for interval in intervals:
            if answer and answer[-1].end >= interval.start:
                if answer[-1].end <= interval.end:
                    answer[-1].end = interval.end
            else:
                answer.append(interval)
        return answer
        

File: apple/LC56MergeIntervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        Collections.sort(intervals, new Comparator<Interval>() {
            public int compare(Interval obj0, Interval obj1) {
                return obj0.start - obj1.start;
            }
        });
        List<Interval> answer = new ArrayList<>();
        Interval prev = null;
        for (Interval inter:intervals) {
            if (prev == null || inter.start > prev.end) {
                answer.add(inter);
                prev = inter;
            }
            else if (prev.end < inter.end){
                prev.end = inter.end;
            }
        }
        return answer;
    }
}
